### PR TITLE
add k8s secret auth method

### LINKED ISSUE #35: Support provide kubeconfig to fetch private registry auth creds from k8s secret
Currently, nydus snapshotter only supports fetching private registry auth creds from docker config.json. But in k8s world, there is usually a [secret](https://kubernetes.io/docs/tasks/configure-pod-container/pull-image-private-registry/) of type `kubernetes.io/dockerconfigjson`, we can also fetch the creds from the secret by providing a kubeconfig to snapshotter.

### PR DESCRIPTION
**Pull request by raoxiang1996**
use k8s informer to watch secret to get registry auth
fix #35 
Signed-off-by: raoxiang raoxiang1208@qq.com

### CODE REVIEW

File: cmd/containerd-nydus-grpc/pkg/command/flags.go [hunk 1]
```diff
@@ -202,6 +203,12 @@ func buildFlags(args *Args) []cli.Flag {
 			Usage:       "whether to validate integrity of image bootstrap",
 			Destination: &args.ValidateSignature,
 		},
+		&cli.StringFlag{
+			Name:        "kubeconfig-path",
+			Value:       "",
```
Comment (imeoer) [inline cmd/containerd-nydus-grpc/pkg/command/flags.go:208]: Maybe provide a default path, like `$KUBECONFIG` or `~/.kube/config`.

File: go.mod [hunk 2]
```diff
@@ -1,6 +1,6 @@
 module github.com/containerd/nydus-snapshotter
 
-go 1.17
+go 1.18
```
Comment (imeoer) [inline go.mod:3]: Why upgrade to `1.18`?

File: cmd/containerd-nydus-grpc/pkg/command/flags.go (hunk 1 above)
Comment (raoxiang1996) [reply cmd/containerd-nydus-grpc/pkg/command/flags.go:208]: if path is empty，NewDefaultClientConfigLoadingRules will default to use ~/.kube/config or $KUBECONFIG

File: go.mod (hunk 2 above)
Comment (raoxiang1996) [reply go.mod:3]: sorry, my go environment is used by default

**Review by imeoer — APPROVED** [review]
LGTM!

File: pkg/auth/kubesecret.go [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+var kubeSecretListener *KubeSecretListener
+
+type KubeSecretListener struct {
+	dockerConfigs map[string]*configfile.ConfigFile
+	configMu      sync.Mutex
+	informer      cache.SharedIndexInformer
+}
+
+func InitKubeSecretListener(ctx context.Context, kubeconfigPath string) {
+	if kubeSecretListener != nil {
```
Comment (changweige) [inline pkg/auth/kubesecret.go:32]: This is still not concurrency-safe.  It is possible to 2 go routines concurrently initialize `kubeSecretListener`. Need to synchronize them?

File: cmd/containerd-nydus-grpc/app/snapshotter/main.go [hunk 4]
```diff
@@ -28,5 +29,6 @@ func Start(ctx context.Context, cfg config.Config) error {
 	opt := ServeOptions{
 		ListeningSocketPath: cfg.Address,
 	}
+	auth.InitKubeSecretListener(ctx, cfg.KubeconfigPath)
```
Comment (changweige) [inline cmd/containerd-nydus-grpc/app/snapshotter/main.go:32]: `InitKubeSecretListener` does not tell its failure. Can tolerate its failure even the option `--kubeconfig-path` is given?

File: pkg/auth/kubesecret.go [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+	kubelistener.configMu.Unlock()
+	return nil
+}
+
+func (kubelistener *KubeSecretListener) deleteDockerConfig(key string) error {
+	kubelistener.configMu.Lock()
+	delete(kubelistener.dockerConfigs, key)
+	kubelistener.configMu.Unlock()
+	_, exists := kubelistener.dockerConfigs[key]
+	if exists {
```
Comment (changweige) [inline pkg/auth/kubesecret.go:91]: The double check is not reliable due to possible insertion concurrency. I suppose we should trust Golang Map and no need to check its existence.

File: pkg/auth/kubesecret.go [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+	kubelistener.configMu.Lock()
+	defer kubelistener.configMu.Unlock()
+	for _, dockerConfig := range kubelistener.dockerConfigs {
+		// Find the auth for the host.
+		authConfig, err := dockerConfig.GetAuthConfig(host)
+		if err != nil {
+			logrus.WithError(err).Infof("failed to get auth config for host %s", host)
+			continue
+		}
+		if len(authConfig.Username) != 0 && len(authConfig.Password) != 0 {
```
Comment (changweige) [inline pkg/auth/kubesecret.go:167]: Is it legal that the password is empty? 

```
username: "tom",
password: "",
```

**Review by changweige — CHANGES_REQUESTED** [review]
Otherwise, LGTM

File: pkg/auth/kubesecret.go (hunk 6 above)
Comment (raoxiang1996) [reply pkg/auth/kubesecret.go:167]: --docker-username, --docker-password and --docker-server is required

**Comment by changweige** [comment]
Please rebase to the latest main branch HEAD to pass the smoke test

**Review by changweige — APPROVED** [review]

**Pull request merged**